### PR TITLE
Add support for the custom-page-sizes proposal

### PR DESCRIPTION
**Pull request by keithw**
This PR adds support for the custom-page-sizes proposal to the lexer, the binary/text parsers and writers, the validator, interpreter, and objdump.

Will plan to add wasm2c in a future PR.

(@fitzgen)

### CODE REVIEW

File: include/wabt/interp/interp-inl.h [hunk 1]
```diff
@@ -64,11 +64,16 @@ inline bool MemoryType::classof(const ExternType* type) {
...
-inline MemoryType::MemoryType(Limits limits)
-    : ExternType(ExternKind::Memory), limits(limits) {
+inline MemoryType::MemoryType(Limits limits, uint32_t page_size)
+    : ExternType(ExternKind::Memory), limits(limits), page_size(page_size) {
   // Always set max.
   if (!limits.has_max) {
-    this->limits.max = limits.is_64 ? WABT_MAX_PAGES64 : WABT_MAX_PAGES32;
+    this->limits.max =
+        limits.is_64 ? WABT_MAX_DEFAULT_PAGES64 : WABT_MAX_DEFAULT_PAGES32;
+    if (page_size == 1) {
```
[the remark this answers is not in the document]
Comment (fitzgen) [reply include/wabt/interp/interp-inl.h:73]: > Why call out 1 in particular here? Is the proposal to add only 1 as an alternate page size?

The current proposal only allows for 1 and 16 as the encoded (log2) page size, i.e. a page size of a single byte or the default Wasm page size. This is a concession to practicality of moving through the standards process, but we are intentionally encoding and spec'ing everything such that this restriction could be loosened in the future to allow any power-of-two page size.

I humbly suggest following that lead in implementations as well, and effectively allowing any power-of-two everywhere except for the validator. FWIW, I personally didn't find that this approach imposed any additional complexity to the implementation in Wasmtime.

**Review by fitzgen — COMMENTED** [review]
Very cool! Love to see it! I'll add a link to this PR from the proposal's `Overview.md`.

**Comment by fitzgen** [comment]
FWIW, if you want to hammer on this implementation a bit, `wasm-smith` has support for generating pseudorandom Wasm modules that use the custom-page-sizes proposal and you can use it via the CLI as well if you don't want to link to Rust code:

```
$ head -n 100 < /dev/urandom | wasm-tools smith -t --custom-page-sizes-enabled=true --max-memories=20
(module
  (table (;0;) 567 296841 externref)
  (memory (;0;) i64 345681 (pagesize 0x1))
  (global (;0;) f64 f64.const -0x1.cbe613cfc1799p+665 (;=-275023225219671150000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000;))
  (global (;1;) externref ref.null extern)
  (export "" (table 0))
  (export "\u{1d}k1" (table 0))
  (elem (;0;) declare funcref)
  (data (;0;) "")
)
```

You can throw that in a loop, pass each test case to WABT, and see if you can find any bugs :)

(There are also lots of other knobs you can configure to control Wasm module generation, for example you can disable random imports to make the modules self contained, turn off the GC proposal, etc...)

File: include/wabt/interp/interp-inl.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (keithw) [reply include/wabt/interp/interp-inl.h]: Thanks for the feedback -- I refactored to allow currently-invalid page sizes everywhere except the validator.

File: src/shared-validator.cc [hunk 2]
```diff
@@ -134,13 +134,33 @@ Result SharedValidator::OnTable(const Location& loc,
...
+  }
+
+  uint64_t absolute_max;
+  const char* desc = (page_size == 1) ? "single-byte pages" : "pages";
+  if (page_size == 1) {
+    absolute_max = limits.is_64 ? std::numeric_limits<uint64_t>::max()
+                                : std::numeric_limits<uint32_t>::max();
+  } else {
+    absolute_max =
+        limits.is_64 ? WABT_MAX_DEFAULT_PAGES64 : WABT_MAX_DEFAULT_PAGES32;
```
[the remark this answers is not in the document]
Comment (keithw) [reply src/shared-validator.cc]: Done.

File: src/shared-validator.cc [hunk 3]
```diff
@@ -134,13 +134,33 @@ Result SharedValidator::OnTable(const Location& loc,
...
     result |= PrintError(loc, "only one memory block allowed");
   }
-  result |= CheckLimits(
-      loc, limits, limits.is_64 ? WABT_MAX_PAGES64 : WABT_MAX_PAGES32, "pages");
+
+  if (page_size != WABT_DEFAULT_PAGE_SIZE) {
+    if (!options_.features.custom_page_sizes_enabled()) {
+      result |= PrintError(loc, "only default page size (64 KiB) is allowed");
+    } else if (page_size != 1) {
+      result |= PrintError(loc, "only page sizes of 1 B or 64 KiB are allowed");
```
[the remark this answers is not in the document]
Comment (keithw) [reply src/shared-validator.cc:149]: Yeah, this is a validation condition of the current proposal, so I think we have to enforce this here at least.

**Pull request merged**